python/src/ops.cpp

```diff
@@ -5189,4 +5189,54 @@ void init_ops(nb::module_& m) {
       Returns:
         array: The row or col contiguous output.
     )pbdoc");
+  m.def(
+      "broadcast_shapes",
+      [](const nb::args& shapes) {
+        if (shapes.size() == 0)
+          throw std::invalid_argument(
+              "broadcast_shapes expects a sequence of shapes");
+
+        std::vector<mx::Shape> shape_vec;
+        shape_vec.reserve(shapes.size());
+
+        for (size_t i = 0; i < shapes.size(); ++i) {
+          if (!nb::isinstance<mx::Shape>(shapes[i]) &&
+              !nb::isinstance<nb::tuple>(shapes[i]))
+            throw std::invalid_argument(
+                "broadcast_shapes expects a sequence of shapes (tuple or list of ints)");
+
+          shape_vec.push_back(nb::cast<mx::Shape>(shapes[i]));
```

Comment:
What if we just broadcast here instead of making a temporary vector of shapes?

```suggestion
          auto shape = nb::cast<mx::Shape>(shapes[i]);
          result = i == 0 ? shape : mx::broadcast_shapes(result, shape);
```